### ISSUE TITLE
Fresh installation fails with clickhouse-init job error

### ISSUE THREAD

**Issue by preved911** [sentry-kubernetes/charts#520]
Hello. I'm trying fresh install sentry. It's fails, because `clickhouse-init` job execution unsuccesful. I see this error in pod logs:
```
Received exception from server (version 20.8.9):
Code: 36. DB::Exception: Received from 127.0.0.1:9000. DB::Exception: Engine Distributed doesn't support TTL clause. Currently only the following engines have support for the feature: [MergeTree, ReplicatedVersionedCollapsingMergeTree, ReplacingMergeTree, ReplicatedSummingMergeTree, ReplicatedAggregatingMergeTree, ReplicatedCollapsingMergeTree, ReplicatedGraphiteMergeTree, ReplicatedMergeTree, ReplicatedReplacingMergeTree, VersionedCollapsingMergeTree, SummingMergeTree, GraphiteMergeTree, CollapsingMergeTree, AggregatingMergeTree].
```

**Comment by martin31821** [comment]
Seeing the same issue after trying to upgrade from a 21.8.0 to 21.10.0.

**Comment by ledmonster** [comment]
I got same error and investigated.

### running queries for each table

```
clickhouse-client --user default --password "" --database=default --host=sentry-clickhouse-$i.sentry-clickhouse-headless --port=9000 --query="DROP TABLE IF EXISTS ${tbl}_dist";
clickhouse-client --user default --password "" --database=default --host=sentry-clickhouse-$i.sentry-clickhouse-headless --port=9000 --query="CREATE TABLE ${tbl}_dist AS ${tbl}_local ENGINE = Distributed('s
entry-clickhouse', 'default', ${tbl}_local, rand())";
```

### show tables
```
$ kubectl exec sentry-clickhouse-0 -- clickhouse-client  --user default --password "" --database=default --host=sentry-clickhouse-0.sentry-clickhouse-headless --query "show tables;"

discover_dist
discover_local
...
transactions_local
```
no `transactions_dist` table found. So the job fails on `transactions_dist` table creation.

### reproduce the error

```
$ kubectl exec sentry-clickhouse-0 -- clickhouse-client  --user default --password "" --database=default --host=sentry-clickhouse-0.sentry-clickhouse-headless --query "CREATE TABLE transactions_dist AS transactions_local ENGINE = Distributed('sentry-clickhouse', 'default', transactions_local, rand());"
Received exception from server (version 20.8.9):
Code: 36. DB::Exception: Received from sentry-clickhouse-0.sentry-clickhouse-headless:9000. DB::Exception: Engine Distributed doesn't support TTL clause. Currently only the following engines have support for the feature: [MergeTree, ReplicatedVersionedCollapsingMergeTree, ReplacingMergeTree, ReplicatedSummingMergeTree, ReplicatedAggregatingMergeTree, ReplicatedCollapsingMergeTree, ReplicatedGraphiteMergeTree, ReplicatedMergeTree, ReplicatedReplacingMergeTree, VersionedCollapsingMergeTree, SummingMergeTree, GraphiteMergeTree, CollapsingMergeTree, AggregatingMergeTree]. 
command terminated with exit code 36
```

### error reason

```
$ kubectl exec sentry-clickhouse-0 -- clickhouse-client  --user default --password "" --database=default --host=sentry-clickhouse-0.sentry-clickhouse-headless --query "show create table transactions_local;" | sed 's/\\n/\n/g'
CREATE TABLE default.transactions_local
(
  ...
)
ENGINE = ReplacingMergeTree(deleted)
PARTITION BY (retention_days, toMonday(finish_ts))
ORDER BY (project_id, toStartOfDay(finish_ts), transaction_name, cityHash64(span_id))
SAMPLE BY cityHash64(span_id)
TTL finish_ts + toIntervalDay(retention_days)
SETTINGS index_granularity = 8192
```

It has `TTL` clause, so the job try to create `transactions_dist` using this table as a template and fails.

**Comment by ledmonster** [comment]
I'm not familiar with clickhouse and snuba, and I don't know what should I do for this problem.

**Comment by ledmonster** [comment]
clickhouse used in default `values.yaml` is very old (`20.8.9.6`). Doesn't it affect this issue?

**Comment by preved911** [comment]
Clickhouse version upgrading also doesn't resolve problem. And snuba [doesn't support ](https://github.com/sentry-kubernetes/charts/issues/366) clickhouse 21.x. I created table with query by hand:
```sql
CREATE TABLE default.transactions_dist
(
    `project_id` UInt64,
    `event_id` UUID,
    `trace_id` Nullable(UUID),
    `span_id` UInt64,
    `transaction_name` LowCardinality(String),
    `transaction_hash` UInt64 MATERIALIZED cityHash64(transaction_name),
    `transaction_op` LowCardinality(String),
    `transaction_status` UInt8 DEFAULT 2,
    `start_ts` DateTime,
    `start_ms` UInt16,
    `finish_ts` DateTime,
    `finish_ms` UInt16,
    `duration` UInt32,
    `platform` LowCardinality(String),
    `environment` LowCardinality(Nullable(String)),
    `release` LowCardinality(Nullable(String)),
    `dist` LowCardinality(Nullable(String)),
    `ip_address_v4` Nullable(IPv4),
    `ip_address_v6` Nullable(IPv6),
    `user` String DEFAULT '',
    `user_hash` UInt64 MATERIALIZED cityHash64(user),
    `user_id` Nullable(String),
    `user_name` Nullable(String),
    `user_email` Nullable(String),
    `sdk_name` LowCardinality(String) DEFAULT '',
    `sdk_version` LowCardinality(String) DEFAULT '',
    `http_method` LowCardinality(Nullable(String)),
    `http_referer` Nullable(String),
    `tags.key` Array(String),
    `tags.value` Array(String),
    `_tags_flattened` String,
    `_tags_hash_map` Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(replaceRegexpAll(k, '(\\=|\\\\)', '\\\\\\1'), '=', v)), tags.key, tags.value),
    `contexts.key` Array(String),
    `contexts.value` Array(String),
    `_contexts_flattened` String,
    `measurements.key` Array(LowCardinality(String)),
    `measurements.value` Array(Float64),
    `span_op_breakdowns.key` Array(LowCardinality(String)),
    `span_op_breakdowns.value` Array(Float64),
    `partition` UInt16,
    `offset` UInt64,
    `message_timestamp` DateTime,
    `retention_days` UInt16,
    `deleted` UInt8,
    `type` LowCardinality(String) MATERIALIZED 'transaction',
    `message` String MATERIALIZED transaction_name,
    `title` String MATERIALIZED transaction_name,
    `timestamp` DateTime MATERIALIZED finish_ts
)
ENGINE = Distributed('sentry-clickhouse', 'default', 'transactions_local', rand())
```
Maybe it's acceptable solution?

**Comment by martin31821** [comment]
As far as I can tell, this should be fixed by #510, since it then uses snuba directly to create these tables, but that yielded other errors for me and also won't work for updates. (will report back)

**Comment by ledmonster** [comment]
@martin31821 thank you.

@preved911 I tried to create `transactions_dist` table manually, but got same error...

```
CREATE TABLE default.transactions_dist
(
   ...
)
ENGINE = Distributed('sentry-clickhouse', 'default', 'transactions_local', rand())
...
Received exception from server (version 20.8.9):
Code: 36. DB::Exception: Received from sentry-clickhouse-0.sentry-clickhouse-headless:9000. DB::Exception: Engine Distributed doesn't support TTL clause. Currently only the following engines have support for the feature: [MergeTree, ReplicatedVersionedCollapsingMergeTree, ReplacingMergeTree, ReplicatedSummingMergeTree, ReplicatedAggregatingMergeTree, ReplicatedCollapsingMergeTree, ReplicatedGraphiteMergeTree, ReplicatedMergeTree, ReplicatedReplacingMergeTree, VersionedCollapsingMergeTree, SummingMergeTree, GraphiteMergeTree, CollapsingMergeTree, AggregatingMergeTree]. 
```

**Comment by ledmonster** [comment]
By upgrading to `21.9.0` through [hard stop](https://develop.sentry.dev/self-hosted/releases/#hard-stops) `21.6.3`, "create table" below worked.

```
CREATE TABLE default.transactions_dist
(
    `project_id` UInt64,
    `event_id` UUID,
    `trace_id` Nullable(UUID),
    `span_id` UInt64,
    `transaction_name` LowCardinality(String),
    `transaction_hash` UInt64 MATERIALIZED cityHash64(transaction_name),
    `transaction_op` LowCardinality(String),
    `transaction_status` UInt8 DEFAULT 2,
    `start_ts` DateTime,
    `start_ms` UInt16,
    `finish_ts` DateTime,
    `finish_ms` UInt16,
    `duration` UInt32,
    `platform` LowCardinality(String),
    `environment` LowCardinality(Nullable(String)),
    `release` LowCardinality(Nullable(String)),
    `dist` LowCardinality(Nullable(String)),
    `ip_address_v4` Nullable(IPv4),
    `ip_address_v6` Nullable(IPv6),
    `user` String DEFAULT '',
    `user_hash` UInt64 MATERIALIZED cityHash64(user),
    `user_id` Nullable(String),
    `user_name` Nullable(String),
    `user_email` Nullable(String),
    `sdk_name` LowCardinality(String) DEFAULT '',
    `sdk_version` LowCardinality(String) DEFAULT '',
    `http_method` LowCardinality(Nullable(String)),
    `http_referer` Nullable(String),
    `tags.key` Array(String),
    `tags.value` Array(String),
    `_tags_flattened` String,
    `_tags_hash_map` Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(replaceRegexpAll(k, '(\\=|\\\\)', '\\\\\\1'), '=', v)), tags.key, tags.value),
    `contexts.key` Array(String),
    `contexts.value` Array(String),
    `_contexts_flattened` String,
    `measurements.key` Array(LowCardinality(String)),
    `measurements.value` Array(Float64),
    `span_op_breakdowns.key` Array(LowCardinality(String)),
    `span_op_breakdowns.value` Array(Float64),
    `partition` UInt16,
    `offset` UInt64,
    `message_timestamp` DateTime,
    `retention_days` UInt16,
    `deleted` UInt8,
    `type` LowCardinality(String) MATERIALIZED 'transaction',
    `message` String MATERIALIZED transaction_name,
    `title` String MATERIALIZED transaction_name,
    `timestamp` DateTime MATERIALIZED finish_ts
)
ENGINE = Distributed('gx-sentry-clickhouse', 'default', 'transactions_local', rand())
```

**Comment by walf443** [comment]
This issues comes from this change https://github.com/getsentry/snuba/pull/2105. And this changes included v21.10.0 https://github.com/getsentry/snuba/releases/tag/21.10.0

**Comment by cabrinha** [comment]
> This issues comes from this change [getsentry/snuba#2105](https://github.com/getsentry/snuba/pull/2105). And this changes included v21.10.0 https://github.com/getsentry/snuba/releases/tag/21.10.0

So, is the solution to run clickhouse container version v21.10.0 or to create the tables manually?

**Comment by ledmonster** [comment]
@cabrinha For me, I resolved the issue by creating the table manually.

**Comment by cabrinha** [comment]
> @cabrinha For me, I resolved the issue by creating the table manually.

@ledmonster can you tell me exactly how you created the table?

I tried:

```
kubectl exec -it sentry-clickhouse-0 -- bash

root@sentry-clickhouse-0:/# clickhouse client
ClickHouse client version 20.8.9.6 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.8.9 revision 54438.

sentry-clickhouse :) CREATE TABLE default.transactions_dist
:-] (
:-]     `project_id` UInt64,
:-]     `event_id` UUID,
:-]     `trace_id` Nullable(UUID),
:-]     `span_id` UInt64,
:-]     `transaction_name` LowCardinality(String),
:-]     `transaction_hash` UInt64 MATERIALIZED cityHash64(transaction_name),
:-]     `transaction_op` LowCardinality(String),
:-]     `transaction_status` UInt8 DEFAULT 2,
:-]     `start_ts` DateTime,
:-]     `start_ms` UInt16,
:-]     `finish_ts` DateTime,
:-]     `finish_ms` UInt16,
:-]     `duration` UInt32,
:-]     `platform` LowCardinality(String),
:-]     `environment` LowCardinality(Nullable(String)),
:-]     `release` LowCardinality(Nullable(String)),
:-]     `dist` LowCardinality(Nullable(String)),
:-]     `ip_address_v4` Nullable(IPv4),
:-]     `ip_address_v6` Nullable(IPv6),
:-]     `user` String DEFAULT '',
:-]     `user_hash` UInt64 MATERIALIZED cityHash64(user),
:-]     `user_id` Nullable(String),
:-]     `user_name` Nullable(String),
:-]     `user_email` Nullable(String),
:-]     `sdk_name` LowCardinality(String) DEFAULT '',
:-]     `sdk_version` LowCardinality(String) DEFAULT '',
:-]     `http_method` LowCardinality(Nullable(String)),
:-]     `http_referer` Nullable(String),
:-]     `tags.key` Array(String),
:-]     `tags.value` Array(String),
:-]     `_tags_flattened` String,
:-]     `_tags_hash_map` Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(replaceRegexpAll(k, '(\\=|\\\\)', '\\\\\\1'), '=', v)), tags.key, tags.value),
:-]     `contexts.key` Array(String),
:-]     `contexts.value` Array(String),
:-]     `_contexts_flattened` String,
:-]     `measurements.key` Array(LowCardinality(String)),
:-]     `measurements.value` Array(Float64),
:-]     `span_op_breakdowns.key` Array(LowCardinality(String)),
:-]     `span_op_breakdowns.value` Array(Float64),
:-]     `partition` UInt16,
:-]     `offset` UInt64,
:-]     `message_timestamp` DateTime,
:-]     `retention_days` UInt16,
:-]     `deleted` UInt8,
:-]     `type` LowCardinality(String) MATERIALIZED 'transaction',
:-]     `message` String MATERIALIZED transaction_name,
:-]     `title` String MATERIALIZED transaction_name,
:-]     `timestamp` DateTime MATERIALIZED finish_ts
:-] )
:-] ENGINE = Distributed('sentry-clickhouse', 'default', 'transactions_local', rand())

Ok.

0 rows in set. Elapsed: 0.017 sec.

sentry-clickhouse :) Bye.
```

But, the `sentry-snuba-transactions-consumer` pod is still crashing with:

```
2021-12-02 21:10:21,576 New partitions assigned: {Partition(topic=Topic(name='events'), index=0): 0}
2021-12-02 21:10:23,016 Caught {
  "__type__": "SerializableException",
  "__name__": "ClickhouseWriterError",
  "__message__": "Table default.transactions_dist doesn't exist. (version 20.8.9.6 (official build))",
  "__should_report__": true,
  "__extra_data__": {
    "code": 60,
    "row": null
  }
}, shutting down...
Traceback (most recent call last):
  File "/usr/local/bin/snuba", line 33, in <module>
    sys.exit(load_entry_point('snuba', 'console_scripts', 'snuba')())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/src/snuba/snuba/cli/consumer.py", line 172, in consumer
    consumer.run()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/processor.py", line 108, in run
    self._run_once()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/processor.py", line 143, in _run_once
    self.__processing_strategy.poll()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/transform.py", line 58, in poll
    self.__next_step.poll()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/collect.py", line 123, in poll
    self.__close_and_reset_batch()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/collect.py", line 105, in __close_and_reset_batch
    self.__batch.close()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/collect.py", line 64, in close
    self.__step.close()
  File "/usr/src/snuba/snuba/consumers/consumer.py", line 223, in close
    self.__insert_batch_writer.close()
  File "/usr/src/snuba/snuba/consumers/consumer.py", line 88, in close
    self.__writer.write(
  File "/usr/src/snuba/snuba/clickhouse/http.py", line 244, in write
    batch.join()
  File "/usr/src/snuba/snuba/clickhouse/http.py", line 192, in join
    raise ClickhouseWriterError(message, code=code, row=row)
snuba.clickhouse.errors.ClickhouseWriterError: Table default.transactions_dist doesn't exist. (version 20.8.9.6 (official build))
```

What am I missing here?

I also have "outcomes" and "sessions" missing as well... Are all the table schemas the same?

**Comment by DaspawnW** [comment]
I run into exactly the same exception when I tried to upgrade without seeing this issue and was surprised that there is no newer version available. I directly upgraded to 22.1.0 which was maybe not the best way to go :-D
But at the end I got it running by simply connecting to clickhouse, taking the ddl of transactions_local table and rename the table to transactions_dist.
Now it works again like a charm. Here the commands for me:

```
kubectl exec -it sentry-clickhouse-0 bash

clickhouse-client  --user default --password "" --database=default --host=sentry-clickhouse-0.sentry-clickhouse-headless

### list available tables
show tables;

### drop table transactions_dist if exists
drop table transactions_dist;

### load information about transactions_local table
show create table transactions_local;

### replace the table name within the create table statement and execute it
CREATE TABLE default.transactions_dist ....
```

**Comment by masterkain** [comment]
I can't get this to work, even if I recreate `default.transactions_dist` manually the clickhouse init job wants to drop and create the table, which is not working.

```
│       echo "clickhouse-init started"                                                                                                                                                                                                                                                   │
│                                                                                                                                                                                                                                                                                        │
│       for tbl in discover errors groupassignee groupedmessage outcomes_hourly migrations outcomes_mv_hourly outcomes_raw sentry sessions_hourly sessions_hourly_mv sessions_raw transactions; do                                                                                       │
│         for ((i=0;i<3;i++)); do                                                                                                                                                                                                                                                        │
│           check_readiness "sentry-clickhouse-$i.sentry-clickhouse-headless" "9000" "5";                                                                                                                                                                                                │
│                                                                                                                                                                                                                                                                                        │
│           clickhouse-client --user default --password "" --database=default --host=sentry-clickhouse-$i.sentry-clickhouse-headless --port=9000 --query="DROP TABLE IF EXISTS ${tbl}_dist";                                                                                             │
│           clickhouse-client --user default --password "" --database=default --host=sentry-clickhouse-$i.sentry-clickhouse-headless --port=9000 --query="CREATE TABLE ${tbl}_dist AS ${tbl}_local ENGINE = Distributed('sentry-clickhouse', 'default', ${tbl}_local, rand())";          │
│         done                                                                                                                                                                                                                                                                           │
│       done                                                                                                                                                                                                                                                                             │
│                                                                                                                                                                                                                                                                                        │
│       echo "clickhouse-init finished"
```

**Comment by masterkain** [comment]
ended up using a custom helm chart https://github.com/sentry-kubernetes/charts/pull/521
but this is (and has been) way too painful

**Comment by Mokto** [comment]
Should be fixed in Release 14.0.0.

Please reopen if needed.
